ipld/hamt/src/node.rs

```diff
@@ -174,6 +174,109 @@ where
         Ok(())
     }
 
+    pub(crate) fn for_each_ranged<Q: ?Sized, S, F>(
+        &self,
+        store: &S,
+        conf: &Config,
+        mut starting_cursor: Option<(HashBits, &Q)>,
+        limit: Option<usize>,
+        f: &mut F,
+    ) -> Result<(usize, Option<K>), Error>
+    where
+        K: Borrow<Q> + Clone,
+        Q: Eq + Hash,
+        F: FnMut(&K, &V) -> anyhow::Result<()>,
+        S: Blockstore,
+    {
+        // determine which subtree the starting_cursor is in
+        let cindex = match starting_cursor {
+            Some((ref mut bits, _)) => {
+                let idx = bits.next(conf.bit_width)?;
+                self.index_for_bit_pos(idx)
+            }
+            None => 0,
+        };
+
+        let mut traversed_count = 0;
+
+        // skip exploration of subtrees that are before the subtree which contains the cursor
+        for p in &self.pointers[cindex..] {
+            match p {
+                Pointer::Link { cid, cache } => {
+                    if let Some(cached_node) = cache.get() {
+                        let (traversed, key) = cached_node.for_each_ranged(
+                            store,
+                            conf,
+                            starting_cursor.take(),
+                            limit.map(|l| l.checked_sub(traversed_count).unwrap()),
+                            f,
+                        )?;
+                        traversed_count += traversed;
+                        if limit.map_or(false, |l| traversed_count > l) && key.is_some() {
```

Comment:
```suggestion
                        if limit.map_or(false, |l| traversed_count >= l) && key.is_some() {
```

(I think?)

Reply:
I think this is a difference now that the cursor refers to the next node to traverse, rather than previous. This boundary should be exercised by a test.

Reply:
Ah, I see. Yeah, makes sense.